bigrig/config.py

```diff
@@ -0,0 +1,229 @@
+
+        try:
+            with open(path, "rt") as fid:
+                credentials = yaml.safe_load(fid)
+        except yaml.YAMLError:
+            # Raise from None to avoid accidental leaks
+            raise yaml.YAMLError(
+                f"Unable to load credentials from `{path}`, it appears to be a non-yaml file"
+            ) from None
+        except OSError as exc:
+            raise OSError(f"Unable to load credentials from `{path}`") from exc
+
+        # As these are loaded later, a runtime validation is needed
+        jsonschema.validate(instance=credentials, schema=CREDENTIALS_SCHEMA)
+
+        return cls(username=credentials["username"], password=credentials["password"])
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}(username='{self.username}', password='***')"
+
+
+@dataclass
+class Source:
+    location: str
+    credentials: t.Optional["Credentials"]
+
+    @classmethod
+    def from_dict(cls, blob: t.Dict) -> "Source":
+        return cls(
+            location=blob["location"],
+            credentials=Credentials.from_path(path=blob.get("credentialsPath")),
+        )
+
+
+@dataclass
+class Target:
+    location: str
+    variables: t.Dict[str, t.Any]
+    credentials: t.Optional["Credentials"]
+
+    @classmethod
+    def from_dict(cls, blob: t.Dict) -> "Target":
+        return cls(
+            location=blob["location"],
+            variables=blob["variables"],
+            credentials=Credentials.from_path(path=blob.get("credentialsPath")),
+        )
+
+
+@dataclass
+class ConfigEntry:
+    origin: Origin
+    source: Source
+    targets: t.Dict[str, Target]
+
+    @classmethod
+    def from_dict(cls, blob: t.Dict) -> "ConfigEntry":
+        jsonschema.validate(instance=blob, schema=CONFIG_SCHEMA)
+        return cls(
+            origin=Origin.from_dict(blob["origin"]),
+            source=Source.from_dict(blob["source"]),
+            targets={
+                name: Target.from_dict(target)
+                for name, target in blob["targets"].items()
+            },
+        )
+
+
+@dataclass
+class RootConfig:
+    entries: t.List[ConfigEntry]
+    packages: t.List[Requirement]
+
+    @classmethod
+    def get_instance(cls) -> "RootConfig":
+        config_path = os.environ.get("BIGRIG_CONFIG_PATH")
+        if not config_path:
+            raise ValueError(
+                f"Missing path to bigrig configuration, point BIGRIG_CONFIG_PATH to bigrig"
+                f"configuration file"
+            )
+
+        packages_path = os.environ.get("BIGRIG_PACKAGES_PATH")
+        if not packages_path:
+            raise ValueError(
+                f"Missing path to bigrig package list, point BIGRIG_PACKAGES_PATH to bigrig"
+                f"package file list"
+            )
+
+        try:
+            with open(config_path, "rt") as fid:
+                entries = [
+                    ConfigEntry.from_dict(blob=blob) for blob in yaml.safe_load_all(fid)
```

Comment:
We can have more then one?

Reply:
As you proposed, yes. And I would like to point out that allowing for more than one config object at a time, lets us have multiple origins to multiple targets, etc in a fairly trivial way so I'd leave it as is.

Reply:
Did I? :)
So you mean going through configs as a list and run for each of them?

Reply:
yea, this is related to my questions regarding packages per origin and what not :)

Reply:
removed as agreed on slack, multiple origin/target will be handled by multiple instances of the app ( for now )